### PR TITLE
Indexes in flush, direct read & write

### PR DESCRIPTION
**Pull request by e5l**

### CODE REVIEW

File: core/commonMain/src/kotlinx/io/Output.kt [hunk 1]
```diff
@@ -113,32 +86,47 @@ abstract class Output(bufferSize: Int = DEFAULT_BUFFER_SIZE) : Closeable {
...
         position = newPosition
+        return result
+    }
+
+    internal fun size(): Int {
+        return position
     }
 
-    private inline fun writePrimitive(
+    internal inline fun writePrimitive(
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Output.kt]: Same here about inlining

File: core/commonMain/src/kotlinx/io/Input.kt [hunk 2]
```diff
@@ -273,50 +334,19 @@ public abstract class Input : Closeable {
...
-            endOffset = limit
-        }
-        val newPosition = reader(buffer, startOffset, endOffset)
-        position = newPosition
-    }
+    protected abstract fun closeSource()
 
     /**
      * Reads a primitive of [primitiveSize] either directly from a buffer with [readDirect]
      * or byte-by-byte into a Long and using [fromLong] to convert to target type.
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Input.kt]: `fromLong` does not longer resolves

File: core/commonMain/src/kotlinx/io/Input.kt [hunk 3]
```diff
@@ -340,93 +371,178 @@ public abstract class Input : Closeable {
...
+
+        return result
     }
 
     /**
      * Reads [size] unsigned bytes from an Input and calls [consumer] on each of them.
      * @throws EOFException if no more bytes available.
      */
-    private inline fun readBytes(size: Int, consumer: (unsignedByte: Int) -> Unit) {
+    private fun readBytes(size: Int, consumer: (unsignedByte: Int) -> Unit) {
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Input.kt]: This method is no longer inline, but still accepts the lambda that manipulates with primitive.
It has the only caller, so please just `comsumer` logic from a call site into this method

File: core/commonMain/src/kotlinx/io/Input.kt [hunk 4]
```diff
@@ -340,93 +371,178 @@ public abstract class Input : Closeable {
...
-                it.append(buffer, limit)
-            }
+        val newPosition = reader(buffer, position, limit)
+        val result = newPosition - position
+        position = newPosition
+        return result
+    }
+
+    /**
+     * Copy buffered bytes to [destination]. If no bytes buffered it will block until [fill] is finished.
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Input.kt]: `Copies currently buffered bytes to the given [destination]. If no bytes are currently buffered, [fill] method is invoked first`

File: core/commonMain/src/kotlinx/io/Input.kt [hunk 5]
```diff
@@ -340,93 +371,178 @@ public abstract class Input : Closeable {
...
+     *
+     * @return transferred bytes count.
+     */
+    private fun copyAvailableTo(destination: Output): Int = destination.writeBuffer { buffer, startIndex, endIndex ->
+        startIndex + copyAvailableTo(buffer, startIndex, endIndex)
+    }
+
+    /**
+     * Copy buffered bytes to [destination]. If no bytes buffered it will call [fill] on [destination] directly.
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Input.kt]: Same here

File: core/commonMain/src/kotlinx/io/InputOperations.kt [hunk 6]
```diff
@@ -1,76 +1,211 @@
 package kotlinx.io
 
 import kotlinx.io.buffer.*
+import kotlin.math.*
+
+/**
+ * Copy [Input] content to [destination].
+ *
+ * The [Input] wouldn't be closed at the end.
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/InputOperations.kt]: The [Input] is not closed at the end of the copy

File: core/commonMain/src/kotlinx/io/InputOperations.kt [hunk 7]
```diff
@@ -1,22 +1,109 @@
...
 
 import kotlinx.io.buffer.*
+import kotlin.math.*
+
+/**
+ * Copy [Input] content to [destination].
+ *
+ * @return transferred bytes count.
+ */
+public fun Input.copyTo(destination: Output): Int {
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply core/commonMain/src/kotlinx/io/InputOperations.kt]: You resolved this conversation but didn't add the comment why it cannot be re-written this way

File: core/commonMain/src/kotlinx/io/InputOperations.kt [hunk 8]
```diff
@@ -1,76 +1,211 @@
...
+
+        count += chunkSize
+    }
+
+    return count
+}
+
+/**
+ *
+ * Copy [size] bytes from [Input] to [destination].
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/InputOperations.kt]: Copies .. from **the** Input to **the** given destination

File: core/commonMain/src/kotlinx/io/InputOperations.kt [hunk 9]
```diff
@@ -1,76 +1,211 @@
...
+        }
+
+        remaining -= chunkSize
+    }
+
+    return size
+}
+
+/**
+ * Read [ByteArray] of fixed [size] from [Input].
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/InputOperations.kt]: Reads

File: core/commonMain/src/kotlinx/io/Output.kt [hunk 10]
```diff
@@ -49,50 +44,30 @@ abstract class Output(bufferSize: Int = DEFAULT_BUFFER_SIZE) : Closeable {
     }
 
     /**
-     * Write a [value] to this [Input].
+     * Bypass [data] from [startOffset] to [endOffset] by using [Output.flush].
+     * If [Output] is not empty, all data will flushed beforehand.
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Output.kt]: will flushed -> will be flushed

File: core/commonMain/src/kotlinx/io/Output.kt [hunk 11]
```diff
@@ -113,32 +88,47 @@ abstract class Output(bufferSize: Int = DEFAULT_BUFFER_SIZE) : Closeable {
     }
 
     /**
-     * Write [source] buffer to destination.
+     * Request to fill [buffer] from [startIndex] to [endIndex] exclusive To destination.
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Output.kt]: It does not fill the buffer

File: core/commonMain/src/kotlinx/io/Output.kt [hunk 12]
```diff
@@ -113,32 +88,47 @@ abstract class Output(bufferSize: Int = DEFAULT_BUFFER_SIZE) : Closeable {
...
+
+    internal fun size(): Int {
+        return position
     }
 
-    private inline fun writePrimitive(
+    internal fun writePrimitive(
         primitiveSize: Int,
         writeDirect: (buffer: Buffer, offset: Int) -> Unit,
         longValue: () -> Long
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Output.kt]: We should get rid of this lambda on the hot path of writing a primitive

File: core/jvmMain/src/kotlinx/io/buffer/BufferOperations.kt [hunk 13]
```diff
@@ -159,3 +158,23 @@ actual inline fun <R> ByteArray.useBuffer(offset: Int, length: Int, block: (Buff
             .slice().order(ByteOrder.BIG_ENDIAN)
     ).let(block)
 }
+
+/**
+ * Compact [Buffer].
+ * Move content from [startIndex] to [endIndex] exclusive to beginning of the buffer.
```
Comment (qwwdfsad) [inline core/jvmMain/src/kotlinx/io/buffer/BufferOperations.kt]: Does it support overlapping?

File: core/commonMain/src/kotlinx/io/text/TextOutputObsolete.kt [hunk 14]
```diff
@@ -11,16 +10,16 @@ private const val lastASCII = 0x7F.toChar()
...
-fun Output.writeUtf8String(text: CharSequence, index: Int = 0, length: Int = text.length - index) {
+public fun Output.writeUtf8String(text: CharSequence, index: Int = 0, length: Int = text.length - index) {
     var textIndex = index // index in text
     val textEndIndex = index + length // index of char after last
     var bytes = 0 // current left bytes to write (in inverted order)
 
     while (textIndex < textEndIndex || bytes != 0) {
-        writeBufferRange { buffer, startOffset, endOffset ->
-            var offset = startOffset
+        writeBuffer { buffer, startIndex, endIndex ->
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/text/TextOutputObsolete.kt]: This change should be in a standalone commit, with corresponding tests and explanation

File: core/commonMain/src/kotlinx/io/Input.kt (hunk 3 above)
Comment (e5l) [reply core/commonMain/src/kotlinx/io/Input.kt]: I'll revert the `inline` modifier for methods with `block`. It was for measure purpose only

File: core/commonMain/src/kotlinx/io/InputOperations.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (e5l) [reply core/commonMain/src/kotlinx/io/InputOperations.kt]: done

File: core/commonMain/src/kotlinx/io/Output.kt (hunk 12 above)
Comment (e5l) [reply core/commonMain/src/kotlinx/io/Output.kt]: I'll revert `inline` remove here and keep lambda. I want to measure the experiment with early `flush` and `buffer.write` instead of `byte-by-byte` copy.

File: benchmarks/commonMain/src/kotlinx/io/benchmarks/InputReadingBenchmark.kt [hunk 15]
```diff
@@ -12,9 +12,9 @@ class InputReadingBenchmark {
 
             override fun closeSource() {}
 
-            override fun fill(buffer: Buffer): Int {
+            override fun fill(buffer: Buffer, startIndex: Int, endIndex: Int): Int {
```
[the remark this answers is not in the document]
Comment (e5l) [reply benchmarks/commonMain/src/kotlinx/io/benchmarks/InputReadingBenchmark.kt]: fixed commit message

File: core/commonMain/src/kotlinx/io/text/TextOutputObsolete.kt (hunk 14 above)
Comment (e5l) [reply core/commonMain/src/kotlinx/io/text/TextOutputObsolete.kt]: This change is code refactoring: migrating to exclusive `endIndex` instead of inclusive `endOffset` and word `Range` remove. It has no logic changes.

File: core/commonMain/src/kotlinx/io/Input.kt [hunk 16]
```diff
@@ -107,95 +78,94 @@ public abstract class Input : Closeable {
...
-                return false
-            bytes.append(buffer, limit)
-            left -= limit
+    /**
+     * Attempts to move chunk from this [Input] to the [destination].
+     */
+    public fun readAvailableTo(
+        destination: Buffer,
+        startIndex: Int = 0,
+        endIndex: Int = destination.size - startIndex
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/Input.kt]: Default value of `endIndex` looks like a bug, it should be `destination.size`.

File: core/jvmMain/src/kotlinx/io/buffer/BufferOperations.kt [hunk 17]
```diff
@@ -159,3 +158,23 @@ actual inline fun <R> ByteArray.useBuffer(offset: Int, length: Int, block: (Buff
...
     ).let(block)
 }
+
+/**
+ * Compacts the [Buffer]. Move content from ([startIndex], [endIndex]) range to (0, 'endIndex - startIndex') range.
+ * The copying ranges can overlap.
+ *
+ * @return [endIndex] - [startIndex] (copied bytes count) or updated [endIndex]
+ */
+internal actual fun Buffer.compact(startIndex: Int, endIndex: Int): Int {
```
Comment (qwwdfsad) [inline core/jvmMain/src/kotlinx/io/buffer/BufferOperations.kt]: Could you please elaborate what's the point of return value?
Why is it useful?

Comment (e5l) [reply core/jvmMain/src/kotlinx/io/buffer/BufferOperations.kt]: It's the new end index, it's easy to combine:
```kotlin
endIndex = buffer.compact(startIndex, endIndex)
```
instead of:
```kotlin
buffer.compact(startIndex, endIndex)
endIndex -= startIndex
```

Comment (qwwdfsad) [reply core/jvmMain/src/kotlinx/io/buffer/BufferOperations.kt]: Good for internal use
